core/object/worker_thread_pool.cpp

```diff
@@ -402,7 +402,9 @@ void WorkerThreadPool::wait_for_group_task_completion(GroupID p_group) {
 		}
 	}
 
-	groups.erase(p_group); // Threads do not access this, so safe to erase here.
+	task_mutex.lock(); // This mutex is needed when Physics 2D and/or 3D is selected to run on a separate thread.
+	groups.erase(p_group);
+	task_mutex.unlock();
```

Comment:
I think performance could be improved if that `groups.erase(p_group);` is moved within the matching `task_mutex.lock()` : this will remove possible overhead from doing two locks in a row.
```cpp
	if (group->low_priority_native_tasks.size() > 0) {
		...
		task_mutex.lock();
		group_allocator.free(group);
		groups.erase(p_group);
		task_mutex.unlock();
	} else {
		...
		if (finished_users == max_users) {
			// All tasks using this group are gone (finished before the group), so clear the group too.
			task_mutex.lock();
			group_allocator.free(group);
			groups.erase(p_group);
			task_mutex.unlock();
		} else {
			task_mutex.lock(); // This mutex is needed when Physics 2D and/or 3D is selected to run on a separate thread.
			groups.erase(p_group);
			task_mutex.unlock();
		}
	}
```

Actually... I wonder if the *intent* of the original code might have been that `groups.erase(p_group);` should only happen in the cases where `group_allocator.free(group)` was called, that is to say `if (group->low_priority_native_tasks.size() > 0)` in the original case, or `if (finished_users == max_users)`  in the else. If this is true, I might suggest the following code instead.
```cpp
	if (group->low_priority_native_tasks.size() > 0) {
		...
		task_mutex.lock();
		group_allocator.free(group);
		groups.erase(p_group);
		task_mutex.unlock();
	} else {
		...
		if (finished_users == max_users) {
			// All tasks using this group are gone (finished before the group), so clear the group too.
			task_mutex.lock();
			group_allocator.free(group);
			groups.erase(p_group);
			task_mutex.unlock();
		}
	}
```